File: hashcode2020_qualification_task/src/main/java/ProblemParser.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class ProblemParser {
    public Problem parse(File fileIn) throws IOException {
        Problem problem = new Problem();

        try (BufferedReader reader = new BufferedReader(new FileReader(fileIn))) {
            String row1 = reader.readLine();
            String[] tokens1 = row1.split(" ");
            problem.B = Integer.parseInt(tokens1[0]);
            problem.L = Integer.parseInt(tokens1[1]);
            problem.D = Integer.parseInt(tokens1[2]);
            problem.books.clear();
            problem.libraries.clear();

            String booksRow = reader.readLine();
            String[] bookScores = booksRow.split(" ");
            for (int b = 0; b < problem.B; b++) {
                problem.add(new Book(b, Integer.parseInt(bookScores[b])));
            }

            for (int l = 0; l < problem.L; l++) {
                String l1 = reader.readLine();
                String l2 = reader.readLine();

                String[] l1Tokens = l1.split(" ");
                String[] l2Tokens = l2.split(" ");

                Library library = new Library(l);
                library.N = Integer.parseInt(l1Tokens[0]);
                library.T = Integer.parseInt(l1Tokens[1]);
                library.M = Integer.parseInt(l1Tokens[2]);

                for (int n = 0; n < library.N; n++) {
                    int bid = Integer.parseInt(l2Tokens[n]);
                    Book book = problem.books.get(bid);
                    library.add(book);
                }

                problem.libraries.add(library);
            }
        }

        return problem;
    }
}

File: cars.2021/src/main/java/ProblemParser.kt
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException

class ProblemParser {
    @Throws(IOException::class)
    fun parse(fileIn: File): Problem {
        val problem = Problem()

        BufferedReader(FileReader(fileIn)).use { reader ->
            val row1 = reader.readLine()
            val row1Tokens = row1.split(" ").toTypedArray()
            problem.simulationDuration = row1Tokens[0].toInt()
            problem.numberOfIntersections = row1Tokens[1].toInt()
            val streetCount = row1Tokens[2].toInt()
            val carCount = row1Tokens[3].toInt()
            problem.bonus = row1Tokens[4].toInt()

            var row: String
            var rowTokens: List<String>
            val streets = HashMap<String, Street>()

            var streetId = 0
            for (i in 0 until streetCount) {
                row = reader.readLine()
                rowTokens = row.split(" ")
                val streetStart = rowTokens[0].toInt()
                val streetEnd = rowTokens[1].toInt()
                val streetName = rowTokens[2]
                val streetDuration = rowTokens[3].toInt()

                val street = Street(streetId, streetName, streetStart, streetEnd, streetDuration)
                problem.streets.add(street)
                streetId++
                streets[streetName] = street
            }
            for (i in 0 until carCount) {
                row = reader.readLine()
                rowTokens = row.split(" ")

                val p = rowTokens[0].toInt()
                val pathStreets = ArrayList<Street>()
                for (pp in 1..p) {
                    val streetName = rowTokens[pp]
                    val street = streets[streetName]!!
                    pathStreets.add(street)
                }
                val path = Path(i, pathStreets)
                val car = Car(i, path)
                problem.cars.add(car)
            }
        }
        return problem
    }
}